presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -641,148 +640,138 @@ public Duration getTotalCpuTime()
+                        schedulerStats.getSplitsScheduledPerIteration().add(result.getSplitsScheduled());
+                        if (result.getBlockedReason().isPresent()) {
+                            switch (result.getBlockedReason().get()) {
+                                case WRITER_SCALING:
+                                    // no-op
+                                    break;
+                                case WAITING_FOR_SOURCE:
+                                    schedulerStats.getWaitingForSource().update(1);
+                                    break;
+                                case SPLIT_QUEUES_FULL:
+                                    schedulerStats.getSplitQueuesFull().update(1);
+                                    break;
+                                case MIXED_SPLIT_QUEUES_FULL_AND_WAITING_FOR_SOURCE:
+                                    schedulerStats.getMixedSplitQueuesFullAndWaitingForSource().update(1);
+                                    break;
+                                case NO_ACTIVE_DRIVER_GROUP:
+                                    schedulerStats.getNoActiveDriverGroup().update(1);
+                                    break;
+                                default:
+                                    throw new UnsupportedOperationException("Unknown blocked reason: " + result.getBlockedReason().get());
+                            }
+                        }
+                    }
+
+                    // make sure to update stage linkage at least once per loop to catch async state changes (e.g., partial cancel)
+                    boolean stageFinishedExecution = false;
+                    for (SqlStageExecution stage : scheduledStages) {
+                        if (!completedStages.contains(stage.getStageId()) && stage.getState().isDone()) {
+                            stageLinkages.get(stage.getStageId())
+                                    .processScheduleResults(stage.getState(), ImmutableSet.of());
+                            completedStages.add(stage.getStageId());
+                            stageFinishedExecution = true;
+                        }
                     }
-                }
 
-                // wait for a state change and then schedule again
-                if (!blockedStages.isEmpty()) {
-                    try (TimeStat.BlockTimer timer = schedulerStats.getSleepTime().time()) {
-                        tryGetFutureValue(whenAnyComplete(blockedStages), 1, SECONDS);
+                    // if any stage has just finished execution try to pull more sections for scheduling
+                    if (stageFinishedExecution) {
+                        break;
                     }
-                    for (ListenableFuture<?> blockedStage : blockedStages) {
-                        blockedStage.cancel(true);
+
+                    // wait for a state change and then schedule again
+                    if (!blockedStages.isEmpty()) {
```

Comment:
As we discussed, it doesn't make sense to wait if there are *any* blocked stages  for at least one of them to finish because there might still be stages that can be scheduled.

Reply:
I just moved the code as it was. It totally doesn't make sense the way it is right now. I'm going to address that in a separate PR.